docs/developers/style-guide.md

```diff
@@ -2,6 +2,94 @@
 title: Style Guide
 ---
 
-!!! warning
+# Style Guide
 
-    This section is work in progress. Please check back soon for the updated content.
+This section collects general coding style guidelines used in Fast-LLM.
+Following these will ensure a swift reviewing process and will help maintaining consistency and readability.
+Note that while we try to enforce these principles,
+exceptions may be allowed on a case-by-case basis, for example if they noticeably improve readability.
+
+As a general principle, **Fast-LLM prioritize code readability and maintainability** over conciseness,
+coding speed or individual programmer's preferences.
+Most of the style choices below are based on this principle.
+
+## 🎯 Basic Style
+
+Unless otherwise specified, Fast-LLM follows the [PEP 8](https://peps.python.org/pep-0008/) coding style.
+This style (and many other conventions) is enforced with automatic formatting through a [pre-commit](https://pre-commit.com/) git hook.
+
+Make sure these git hooks are installed by running
+```bash
+pip install pre-commit
+pre-commit install
+```
+
+!!! note "More on automated formatting"
+
+    Fast-LLM's automated formatting includes [Black](https://black.readthedocs.io/en/stable/),
+    [isort](https://pycqa.github.io/isort/), [autoflake](https://github.com/PyCQA/autoflake), and a few other packages.
+    See Fast-LLM's [pre-commit configuration](https://github.com/ServiceNow/Fast-LLM/blob/main/.pre-commit-config.yaml) for more details.
+
+## 📚 Naming Conventions
+
+Names should be as self-explanatory as possible, within reason.
+This includes python identifiers (classes, variables, methods, modules, etc.), file names and configuration parameters.
+For example:
+
+* Avoid abbreviations, especially domain-specific ones. Ex. `bs -> batch_size`.
+* Avoid meaningless variable names such as `x`, except for generic methods.
+* Avoid redundancies especially for configuration parameters, ex. `data.data_type` -> `data.type`.
+* Avoid name parts that refer to the data type, ex. `num`. Use type hints instead.
+
+Note that these conventions are enforced more strictly on user-facing names since they are more difficult to change,
+for example configuration parameters and the public interface of core classes and modules.
+
+## 🛬 Imports
+
+We use the following conventions for imports (other than those enforced by isort):
+
+* Import standard library and third party modules by module (ex. `import package.module`, not `from package.module import method`).
+This keeps identifier's origin explicit.
+* Avoid renaming with `as`, except for some (arbitrarily chosen) common ones: `numpy as np`, `triton.language as tl`.
+* Import first-party modules through specific identifiers (ex. `from fast_llm.module import method`, not `import fast_llm.module`). This facilitates the tracking of where those objects are used.
```

Comment:
Pycharm isn't clever enough to figure out symbol usage otherwise?

Reply:
Pycharm and other IDEs are, but not humans.

Reply:
so you're saying that you use full-text search over the codebase for the method name to figure out where they are referenced? I don't quite follow this otherwise.

Reply:
My point is we can see what exactly is used in a file just by looking at the top of it.